File: src/java/org/voegtle/weatherstation/server/parser/DataLine.java
package org.voegtle.weatherstation.server.parser;

import java.util.ArrayList;
import java.util.StringTokenizer;

public class DataLine {
  private final ArrayList<String> values = new ArrayList<>();

  public DataLine(String data) {
    data = data.replace(" ", "+");
    data = data.replace(";;", "; ;");
    data = data.replace(";;", "; ;");
    StringTokenizer tokenizer = new StringTokenizer(data, ";\n");
    while (tokenizer.hasMoreTokens()) {
      values.add(tokenizer.nextToken().trim());
    }
  }

  public String get(int index) {
    if (index >= values.size()) {
      return null;
    }
    return values.get(index);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    int i = 0;
    for (String val : values) {
      sb.append(i++);
      sb.append("=<");
      sb.append(val);
      sb.append("> ");
    }
    return sb.toString();
  }
}

File: src/main/java/org/voegtle/weatherstation/server/parser/DataLine.kt
package org.voegtle.weatherstation.server.parser

import java.util.ArrayList
import java.util.StringTokenizer

class DataLine(line: String) {
  private val values = ArrayList<String>()

  init {
    val data = line.replace(" ", "+")
        .replace(";;", "; ;")
        .replace(";;", "; ;")
    val tokenizer = StringTokenizer(data, ";\n")
    while (tokenizer.hasMoreTokens()) {
      values.add(tokenizer.nextToken().trim { it <= ' ' })
    }
  }

  operator fun get(index: Int): String? {
    return if (index >= values.size) {
      null
    } else values[index]
  }

  fun size() = values.size

  override fun toString(): String {
    val sb = StringBuilder()
    var i = 0
    for (value in values) {
      sb.append(i++)
      sb.append("=<")
      sb.append(value)
      sb.append("> ")
    }
    return sb.toString()
  }
}
